agent/engine/docker_image_manager.go

```diff
@@ -388,70 +388,87 @@ func (imageManager *dockerImageManager) getNonECSContainerIDs(ctx context.Contex
 			break
 		}
-		seelog.Infof("Removing non-ECS Image: %s", kv.ImageName)
-		err := imageManager.client.RemoveImage(ctx, kv.ImageName, dockerclient.RemoveImageTimeout)
+		if len(image.RepoTags) > 1 {
+			seelog.Debugf("Non-ECS image has more than one tag Image: %s (Tags: %s)", image.ImageID, image.RepoTags)
+			for _, tag := range image.RepoTags {
+				err := imageManager.client.RemoveImage(ctx, tag, dockerclient.RemoveImageTimeout)
+				if err != nil {
+					seelog.Errorf("Error removing RepoTag (ImageID: %s, Tag: %s) %v", image.ImageID, tag, err)
+				}
+			}
+		}
+		seelog.Debugf("Removing non-ECS Image: %s (Tags: %s)", image.ImageID, image.RepoTags)
+		err := imageManager.client.RemoveImage(ctx, image.ImageID, dockerclient.RemoveImageTimeout)
 		if err != nil {
-			seelog.Errorf("Error removing Image %s - %v", kv.ImageName, err)
+			seelog.Errorf("Error removing Image %s (Tags: %s) - %v", image.ImageID, image.RepoTags, err)
 			continue
 		} else {
-			seelog.Infof("Image removed: %s", kv.ImageName)
+			seelog.Infof("Image removed: %s (Tags: %s)", image.ImageID, image.RepoTags)
 			numImagesAlreadyDeleted++
 		}
 	}
 }
 
-func (imageManager *dockerImageManager) getNonECSImageNames(ctx context.Context) []string {
-	response := imageManager.client.ListImages(ctx, dockerclient.ListImagesTimeout)
-	var allImagesNames []string
-	for _, name := range response.RepoTags {
-		allImagesNames = append(allImagesNames, name)
+// getNonECSImages returns type ImageWithSizeID with all fields populated.
+func (imageManager *dockerImageManager) getNonECSImages(ctx context.Context) []ImageWithSizeID {
+	r := imageManager.client.ListImages(ctx, dockerclient.ListImagesTimeout)
+	var allImages []ImageWithSizeID
+	// inspect all images
+	for _, imageID := range r.ImageIDs {
+		resp, err := imageManager.client.InspectImage(imageID)
+		if err != nil {
+			seelog.Errorf("Error inspecting non-ECS image: (ImageID: %s), %s", imageID, err)
+			continue
+		}
+		allImages = append(allImages,
+			ImageWithSizeID{
+				ImageID:  imageID,
+				Size:     resp.Size,
+				RepoTags: resp.RepoTags,
+			})
 	}
-	var ecsImageNames []string
+
+	// get all 'ecs' image IDs
+	var ecsImageIDs []string
 	for _, imageState := range imageManager.getAllImageStates() {
-		for _, imageName := range imageState.Image.Names {
-			ecsImageNames = append(ecsImageNames, imageName)
-		}
+		ecsImageIDs = append(ecsImageIDs, imageState.Image.ImageID)
 	}
 
-	var nonECSImageNames = exclude(allImagesNames, ecsImageNames)
-	return nonECSImageNames
+	// exclude 'ecs' image IDs and image IDs with an explicitly excluded tag
+	var nonECSImages []ImageWithSizeID
+	for _, image := range allImages {
+		// check image ID is not excluded
+		if !isInExclusionList(image.ImageID, ecsImageIDs) {
+			// check image TAG(s) is not excluded
+			if !anyIsInExclusionList(image.RepoTags, imageManager.imageCleanupExclusionList) {
+				nonECSImages = append(nonECSImages, image)
```

Comment:
We can get rid of this nested if statement by using:
if isInExclustionList(){
   continue
}
if !anyIsInExclusionList(){
}
The nested if statement plus negative condition makes a little bit hard to understand what we are trying to do here.  However, this is not a blocker for you to merge the change.

Reply:
will do